### PR TITLE
Dispose all subscriptions when connection broken

### PR DESCRIPTION
**Pull request by Shane32**
This code ensures that all existing subscriptions are disposed when the connection is disposed.  Of course if the shutdown is clean then there should be no difference. Also it prevents additional subscriptions from connecting after the connection is disposed. Also the subscription releases all held objects upon disposal - so in case the IObserverable does not properly release the observer after it is disposed, at least there won't be any held handles to the underlying WebSocket.

### CODE REVIEW

**Comment by sungam3r** [comment]
approvals?

**Comment by Shane32** [comment]
> approvals?

Whoops!  Should be good now.

**Comment by Shane32** [comment]
@sungam3r Merge if it looks good to you.

**Comment by Shane32** [comment]
@sungam3r Ready for review here.

File: src/Transports.Subscriptions.Abstractions/SubscriptionServer.cs [hunk 1]
```diff
@@ -136,5 +137,7 @@ private void LinkToTransportWriter()
             TransportWriter = Transport.Writer;
             _logger.LogDebug("Writer pipeline created");
         }
+
+        public void Dispose() => (Subscriptions as IDisposable)?.Dispose();
```
Comment (sungam3r) [inline src/Transports.Subscriptions.Abstractions/SubscriptionServer.cs:141]: It's not a "DI-way", but OK for now.

File: src/Transports.Subscriptions.WebSockets/WebSocketConnection.cs [hunk 2]
```diff
@@ -26,6 +26,7 @@ public virtual async Task Connect()
 
         public void Dispose()
         {
+            _server.Dispose();
```
Comment (sungam3r) [inline src/Transports.Subscriptions.WebSockets/WebSocketConnection.cs:29]: Does `WebSocketConnection` owns `SubscriptionServer` ? Are you sure?

File: src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs [hunk 3]
```diff
@@ -45,13 +46,18 @@ public SubscriptionManager(IGraphQLExecuter executer, ILoggerFactory loggerFacto
...
+            if (_disposed)
+                throw new ObjectDisposedException(nameof(SubscriptionManager));
 
             var subscription = await ExecuteAsync(id, payload, context).ConfigureAwait(false);
 
             if (subscription == null)
                 return;
 
-            _subscriptions[id] = subscription;
+            if (_disposed)
```
Comment (sungam3r) [inline src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs:57]: Concurrent dispose?

File: src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs [hunk 4]
```diff
@@ -140,5 +146,28 @@ public Task UnsubscribeAsync(string id)
 
             return null;
         }
+
+        public virtual void Dispose()
```
Comment (sungam3r) [inline src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs:150]: Who calls this method?

File: src/Transports.Subscriptions.WebSockets/WebSocketConnection.cs (hunk 2 above)
Comment (Shane32) [reply src/Transports.Subscriptions.WebSockets/WebSocketConnection.cs:29]: Yes. Check the factory. It’s created by the factory and passed to the connection. Perhaps a poor design but that’s what it is in this version.

Comment (Shane32) [reply src/Transports.Subscriptions.WebSockets/WebSocketConnection.cs:29]: (On the road now so I’m going from memory…)

File: src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs (hunk 3 above)
Comment (Shane32) [reply src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs:57]: The connection can disconnect during ExecuteAsync, so another check is needed

File: src/Transports.Subscriptions.Abstractions/SubscriptionServer.cs (hunk 1 above)
Comment (Shane32) [reply src/Transports.Subscriptions.Abstractions/SubscriptionServer.cs:141]: This is calling SubscriptionManager.Dispose if I remember right

**Comment by Shane32** [comment]
Just checked the code:

`WebSocketConnectionFactory.CreateConnection`:
- Creates `WebSocketTransport` and `SubscriptionManager`.
  - The `WebSocketTransport` disposes the underlying WebSocket
  - The `SubscriptionManager` disposes any `Subscription`s it creates in a thread-safe manner.
    - `Subscription` instances dispose the underlying `IDisposable` returned from subscribing
- Creates `SubscriptionServer` which owns (and disposes) the `SubscriptionManager` and references the `WebSocketTransport`.
- Creates `WebSocketConnection` which owns (and disposes) `WebSocketTransport` and `SubscriptionServer`.
- The `WebSocketConnection` is disposed by the `using` statement in the middleware for ASP.Net Core (`GraphQLWebSocketsMiddleware`).

This may sound confusing, but I'm pretty sure it's all correct.

File: src/Transports.Subscriptions.WebSockets/WebSocketConnection.cs (hunk 2 above)
Comment (Shane32) [reply src/Transports.Subscriptions.WebSockets/WebSocketConnection.cs:29]: Double checked: yes it does.

File: src/Transports.Subscriptions.Abstractions/SubscriptionServer.cs (hunk 1 above)
Comment (Shane32) [reply src/Transports.Subscriptions.Abstractions/SubscriptionServer.cs:141]: Confirmed.

File: src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs (hunk 4 above)
Comment (Shane32) [reply src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs:150]: SubscriptionServer.Dispose

File: src/Transports.Subscriptions.WebSockets/WebSocketConnection.cs (hunk 2 above)
Comment (sungam3r) [reply src/Transports.Subscriptions.WebSockets/WebSocketConnection.cs:29]: OK

File: src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs (hunk 3 above)
Comment (sungam3r) [reply src/Transports.Subscriptions.Abstractions/SubscriptionManager.cs:57]: OK. I still do not know this code.

**Review by sungam3r — APPROVED** [review]

**Pull request merged**